File: python/day18.py
import unittest
from utilities import read_lines
from typing import List


def operation_calculate(num_stack: List[int], operation: str):
    if operation == '':
        return
    a = num_stack.pop()
    b = num_stack.pop()

    if operation == '+':
        result = a + b
    else:
        result = a * b

    num_stack.append(result)


def math_calculate_with_parentheses(expr: str, advanced: bool) -> int:
    opens: int = 0
    open_start: int = -1

    for i, c in enumerate(expr):
        if c == '(':
            opens += 1
            if open_start == -1:
                open_start = i
        elif c == ')':
            opens -= 1
            if opens == 0:
                # Calculate the expression inside parentheses first
                # then create calculate a new expression with the stuff
                # inside parentheses replaced by its result
                value = math_calculate(expr[open_start+1:i], advanced)
                return math_calculate(expr[:open_start] + str(value) + expr[i+1:], advanced)


def math_calculate_without_parentheses(expr: str, advanced) -> int:
    curr_num: str = ''
    curr_operation: str = ''
    num_stack: List[int] = []
    operation_stack: List[str] = []

    for c in expr:
        if c.isdigit():
            curr_num += c
        else:
            if curr_num != '':
                num_stack.append(int(curr_num))
            curr_num = ''

            if not advanced or curr_operation == '+':
                operation_calculate(num_stack, curr_operation)
            elif advanced:
                operation_stack.append(curr_operation)
            curr_operation = c

    if curr_num != '':
        num_stack.append(int(curr_num))
    operation_stack.append(curr_operation)
    while len(operation_stack) != 0:
        operation_calculate(num_stack, operation_stack.pop())

    return num_stack[0]


def math_calculate(expr: str, advanced: bool = False) -> int:
    expr: str = expr.replace(' ', '')
    if '(' in expr:
        return math_calculate_with_parentheses(expr, advanced)
    else:
        return math_calculate_without_parentheses(expr, advanced)


def part_1(expressions: List[str]) -> int:
    return sum(map(math_calculate, expressions))


def part_2(expressions: List[str]) -> int:
    return sum([math_calculate(expr, True) for expr in expressions])


class Testing(unittest.TestCase):

    def test_math_calculate(self):
        self.assertEqual(math_calculate("1 + 2 * 3 + 4 * 5 + 6"), 71)
        self.assertEqual(math_calculate("1 + (2 * 3) + (4 * (5 + 6))"), 51)
        self.assertEqual(math_calculate("2 * 3 + (4 * 5)"), 26)
        self.assertEqual(math_calculate("5 + (8 * 3 + 9 + 3 * 4 * 3)"), 437)
        self.assertEqual(math_calculate(
            "5 * 9 * (7 * 3 * 3 + 9 * 3 + (8 + 6 * 4))"), 12240)
        self.assertEqual(math_calculate(
            "((2 + 4 * 9) * (6 + 9 * 8 + 6) + 6) + 2 + 4 * 2"), 13632)

    def test_math_calculate_advanced(self):
        self.assertEqual(math_calculate(
            "1 + 2 * 3 + 4 * 5 + 6", True), 231)
        self.assertEqual(math_calculate(
            "1 + (2 * 3) + (4 * (5 + 6))", True), 51)
        self.assertEqual(math_calculate("2 * 3 + (4 * 5)", True), 46)
        self.assertEqual(math_calculate(
            "5 + (8 * 3 + 9 + 3 * 4 * 3)", True), 1445)
        self.assertEqual(math_calculate(
            "5 * 9 * (7 * 3 * 3 + 9 * 3 + (8 + 6 * 4))", True), 669060)
        self.assertEqual(math_calculate(
            "((2 + 4 * 9) * (6 + 9 * 8 + 6) + 6) + 2 + 4 * 2", True), 23340)
        self.assertEqual(math_calculate(
            "3 + (2 * 2 + (7 * 3) * 2) + 7 + 4 + (2 + 6 * 4 + 9 * 4 * 5)", True), 2186)


if __name__ == '__main__':
    print('Day 18')
    expressions: List[str] = read_lines('../inputs/day18.txt')

    print('\tPart 1: {}'.format(part_1(expressions)))
    print('\tPart 2: {}'.format(part_2(expressions)))
    unittest.main()

File: rust/src/day18.rs
use crate::utilities::read_lines;

#[derive(Default)]
struct Calculator {
    nums: Vec<i64>,
}

impl Calculator {
    fn perform_operation(&mut self, operation: char) {
        if operation == '+' {
            let a = self.nums.pop().unwrap();
            let b = self.nums.pop().unwrap();
            self.nums.push(a + b);
        } else if operation == '*' {
            let a = self.nums.pop().unwrap();
            let b = self.nums.pop().unwrap();
            self.nums.push(a * b);
        }
    }

    fn store(&mut self, num: i64) {
        self.nums.push(num);
    }

    fn get_result(&self) -> i64 {
        self.nums[0]
    }
}

fn calculate_expression(expr: &str, advanced: bool) -> i64 {
    if expr.contains('(') {
        let mut opens = 0;
        let mut start = 0;
        for (i, c) in expr.chars().enumerate() {
            if c == '(' {
                opens += 1;
                if opens == 1 {
                    start = i;
                }
            } else if c == ')' {
                opens -= 1;
                if opens == 0 {
                    let mut new_expr = expr[..start].to_string();
                    let sub_expression_result =
                        calculate_expression(&expr[start + 1..i], advanced).to_string();
                    new_expr.push_str(sub_expression_result.as_str());
                    new_expr.push_str(&expr[i + 1..]);
                    return calculate_expression(new_expr.as_str(), advanced);
                }
            }
        }
        unreachable!("Only reach on invalid expression")
    } else {
        let mut calculator = Calculator::default();
        let mut curr_num = String::new();
        let mut operation = ' ';
        let mut operation_stack = vec![];

        for c in expr.replace(' ', "").chars() {
            if c.is_ascii_digit() {
                curr_num.push(c);
            } else {
                if !curr_num.is_empty() {
                    calculator.store(curr_num.parse().unwrap());
                    curr_num.clear();
                }

                if !advanced || operation == '+' {
                    calculator.perform_operation(operation);
                } else if advanced {
                    operation_stack.push(operation)
                }
                operation = c;
            }
        }

        if !curr_num.is_empty() {
            calculator.store(curr_num.parse().unwrap());
        }

        operation_stack.push(operation);

        while !operation_stack.is_empty() {
            calculator.perform_operation(operation_stack.pop().unwrap());
        }

        calculator.get_result()
    }
}

pub fn main() {
    println!("Day 18");
    let expressions = read_lines("../inputs/day18.txt");

    println!(
        "\tPart 1: {}",
        expressions
            .iter()
            .map(|expr| calculate_expression(expr.as_str(), false))
            .sum::<i64>()
    );
    println!(
        "\tPart 2: {}",
        expressions
            .iter()
            .map(|expr| calculate_expression(expr.as_str(), true))
            .sum::<i64>()
    );
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_calculate_expression() {
        assert_eq!(calculate_expression("1 + 2 * 3 + 4 * 5 + 6", false), 71);
        assert_eq!(
            calculate_expression("1 + (2 * 3) + (4 * (5 + 6))", false),
            51
        );
        assert_eq!(calculate_expression("2 * 3 + (4 * 5)", false), 26);
        assert_eq!(
            calculate_expression("5 + (8 * 3 + 9 + 3 * 4 * 3)", false),
            437
        );
        assert_eq!(
            calculate_expression("5 * 9 * (7 * 3 * 3 + 9 * 3 + (8 + 6 * 4))", false),
            12240
        );
        assert_eq!(
            calculate_expression("((2 + 4 * 9) * (6 + 9 * 8 + 6) + 6) + 2 + 4 * 2", false),
            13632
        );
    }

    #[test]
    fn test_calculate_expression_advanced() {
        assert_eq!(calculate_expression("1 + 2 * 3 + 4 * 5 + 6", true), 231);
        assert_eq!(
            calculate_expression("1 + (2 * 3) + (4 * (5 + 6))", true),
            51
        );
        assert_eq!(calculate_expression("2 * 3 + (4 * 5)", true), 46);
        assert_eq!(
            calculate_expression("5 + (8 * 3 + 9 + 3 * 4 * 3)", true),
            1445
        );
        assert_eq!(
            calculate_expression("5 * 9 * (7 * 3 * 3 + 9 * 3 + (8 + 6 * 4))", true),
            669060
        );
        assert_eq!(
            calculate_expression("((2 + 4 * 9) * (6 + 9 * 8 + 6) + 6) + 2 + 4 * 2", true),
            23340
        );
        assert_eq!(
            calculate_expression(
                "3 + (2 * 2 + (7 * 3) * 2) + 7 + 4 + (2 + 6 * 4 + 9 * 4 * 5)",
                true
            ),
            2186
        );
    }
}
